src/MetamaskInpageProvider.js

```diff
@@ -169,15 +169,22 @@ module.exports = class MetamaskInpageProvider extends SafeEventEmitter {
 
     // json rpc notification listener
     jsonRpcConnection.events.on('notification', (payload) => {
+
       this.emit('data', payload) // deprecated
-      if (payload.method === 'wallet_accountsChanged') {
-        this._handleAccountsChanged(payload.result)
-      } else if (EMITTED_NOTIFICATIONS.includes(payload.method)) {
-        this.emit('notification', payload) // deprecated
+
+      const { method, params, result } = payload
+
+      if (method === 'wallet_accountsChanged') {
+        this._handleAccountsChanged(result)
+      } else if (EMITTED_NOTIFICATIONS.includes(method)) {
+
         this.emit('message', {
-          type: payload.method,
-          data: payload.params,
+          type: method,
+          data: params,
         })
+
+        // deprecated
+        this.emit('notification', (params && params.result) || payload)
```

Comment:
Why the fallback to `payload`?

The old provider seemed to assume that `params.result` was always set for `eth_subscribe` messages. I'm not sure what we should do if it isn't, but emitting nothing seems like a safer choice than emitting the payload, which is a data structure that listeners would be totally unfamiliar with.
```suggestion
        this.emit('notification', params ? params.result : undefined)
```

Reply:
Or maybe this is better

```suggestion
        if (params) {
          this.emit('notification', params.result)
        }
```

Reply:
96ba786